lib/Dialect/Calyx/CalyxOps.cpp

```diff
@@ -1418,7 +1457,8 @@ static void getCellAsmResultNames(OpAsmSetValueNameFn setNameFn, Operation *op,
 /// Determines whether the given direction is valid with the given inputs. The
 /// `isDestination` boolean is used to distinguish whether the value is a source
 /// or a destination.
-static LogicalResult verifyPortDirection(AssignOp op, Value value,
+template <typename T>
+static LogicalResult verifyPortDirection(T op, Value value,
```

Comment:
> It doesn't look like you're using anything from the specific operation type, so this could probably just be `LogicalResult verifyPortDirection(Operation* op, ...` instead of a template.

I use this to verify the connection of the port in the invoke, with this function my first pass parameter is `InvokeOp` and it works fine, I don't think it is necessary to rewrite a function to determine the direction of the port, so I use the original function here.

Reply:
It's not so much that, it's rather than since this function isn't dependent on anything specific from either `InvokeOp` or `AssignOp`, you're solely relying on `Operation*` when you're using this op (calling `op->emitOpError`). In that case, a template function is inappropriate seeing as it will emit two separate functions (once for each template instance) when only one is needed.